### DISCUSSION TITLE
Resolution order is not consistent with python import resolution

### DISCUSSION THREAD

**Discussion by navignaw** [microsoft/pyright#8762] [General]
This is related to https://github.com/microsoft/pyright/issues/6415 and https://github.com/microsoft/pylance-release/discussions/3708#discussioncomment-7536971.

The [resolution order docs](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md) say the following:

> For absolute (non-relative) imports, Pyright employs the following resolution order:
>
> 1. Try to resolve using the stubPath as defined in the stubPath config entry or the python.analysis.stubPath setting.
>
> 2. Try to resolve using code within the workspace.
>   * Try to resolve relative to the root directory of the execution environment. If no execution environments are specified in the config file, use the root of the workspace. For more information about execution environments, refer to the [configuration documentation](https://github.com/microsoft/pyright/blob/ace7edb41e02b5d6b1490977eff45663bbc43a08/docs/configuration.md#execution-environment-options).
>
> 3. Try to resolve using stubs or inlined types found within installed packages. Pyright uses the configured Python environment to determine whether a package has been installed. For more details about how to configure your Python environment for Pyright, see below. If a Python environment is configured, Pyright looks in the lib/site-packages, Lib/site-packages, or python*/site-packages subdirectory. If no site-packages directory can be found, Pyright attempts to run the configured Python interpreter and ask it for its search paths. If no Python environment is configured, Pyright will use the default Python interpreter by invoking python.

We've found this to lead to false positives in our code. Consider this directory structure:

```
src/
  bar.py
  alembic/
    foo.py
```

Inside `bar.py`, we attempt to import `from alembic import op`. alembic is a 3rd-party package that provides its own type stub, e.g. [op.pyi](https://github.com/sqlalchemy/alembic/blob/main/alembic/op.pyi). However, based on the resolution order, pyright is actually attempting to resolve the import with step 2, and throwing an error because `src/alembic/op.py` does not exist in our codebase:

> error: "op" is unknown import symbol (reportAttributeAccessIssue)

This issue goes away when renaming `src/alembic/` to any other directory, like `src/migrations/`. Alternatively, changing the import to `import alembic.op as op` also resolves the error. However, the python code itself executes correctly as written, which leads to confusing behavior.

Is it possible to swap the resolution order for type stubs and code relative to the root of the repo? Or a way to opt out of relative resolution entirely? It seems like an anti-pattern since python already provides the syntax `import .alembic` to import relative packages.

**Comment by erictraut** [comment]
> Is it possible to swap the resolution order for type stubs and code relative to the root of the repo?

This would create compatibility issues for other pyright users who depend on the current import resolution order, so this isn't a change that would make sense. There are good reasons why pyright uses the import resolution order it does. Much of it is dictated by [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering).

You can control pyright's import resolution behavior by specifying the `extraPaths` in your config file. If you don't specify this, pyright defaults to `src`. That default works well for most projects, but it sounds like it doesn't for yours. You can override the default by providing an explicit `extraPaths` entry, even if it is an empty list of directories.

**Reply by navignaw** [reply]
Thanks for the explanation. I totally understand the difficulty changing the behavior due to backwards compatibility, but I'm still a little confused how exactly this is meant to work.

I gave a simplified example with `src/alembic/`, but the actual directory structure of our project is a little more complex; the file that is reporting the error is actually more like:

```
services/foo_service/alembic/versions/migration.py
```

which does not appear to match the default `extraPaths` of `src`. In fact, it appears to be erroring _regardless of where `alembic` is in the source tree_ - even if I rename it to `services/foo_service/blah/bar/baz/alembic/versions/migration.py`, the same issue happens until I rename `alembic/` to something else.

I attempted to set `extraPaths = []` in my `pyproject.toml` but still ran into the same error. Any guidance would be much appreciated!

**Reply by erictraut** [reply]
Can you provide more detailed repro steps or create a minimal example in a publicly-accessible repository? Import resolution is complex, and there are many things that influence it, so I'll probably need to repro it locally to help you further.

**Reply by navignaw** [reply]
Unfortunately it's a bit tricky to reproduce as we have a fairly complex setup involving Bazel. It seems like our team has deprioritized this for now, but if I make any other headway I'll be sure to come back and let you know. Thanks for the help anyway!
